cmake/ActsCompilerOptions.cmake

```diff
@@ -2,10 +2,23 @@
 if (NOT CMAKE_BUILD_TYPE)
   set(CMAKE_BUILD_TYPE RelWithDebInfo CACHE STRING "Build type configuration" FORCE)
   message(STATUS "Setting default build type: ${CMAKE_BUILD_TYPE}")
-endif() 
+endif()
+
+set(extra_flags "-Wall;-Wextra;-Wpedantic;-Wshadow;-Wno-unused-local-typedefs")
+
+# This adds some useful conversion checks like float-to-bool, float-to-int, etc.
+# However, at the moment this is only added to clang builds, since GCC's -Wfloat-conversion
+# is much more aggressive and also triggers on e.g., double-to-float
+if(CMAKE_CXX_COMPILER_ID MATCHES "Clang|AppleClang")
+  list(APPEND extra_flags "-Wfloat-conversion")
+endif()
 
-set(ACTS_CXX_FLAGS "-Wall -Wextra -Wpedantic -Wshadow -Wno-unused-local-typedefs")
-set(ACTS_CXX_FLAGS_DEBUG "--coverage")
+string(REPLACE " " ";" user_flags "${ACTS_CXX_FLAGS}")
+
+list(APPEND extra_flags ${user_flags})
+
+string(REPLACE ";" " " ACTS_CXX_FLAGS "${extra_flags}")
```

Comment:
This code, I believe, will keep the `ACTS_CXX_FLAGS` variable growing at each re-configuration. :frowning: Remember, since it's coming from the command line, `ACTS_CXX_FLAGS` is a cache variable. It's value is remembered. So when you do an incremental build that requires a re-configuration, all the string manipulation will happen once again on this variable.

Unfortunately there's no perfect solution to this that I would know of. :frowning: I can only suggest to declare the `ACTS_CXX_FLAGS_<FOO>` variables properly as cache variable (otherwise they are really of not much use), and then `FORCE` set `CMAKE_CXX_FLAGS_<FOO>` based on them on every re-configuration with some logic.

Though that is also not super nice when trying to use Acts as a sub-project of something else. :thinking:

Reply:
I guess it depends on what we're actually trying to accomplish. I'm not HUGE fan of the separate `ACTS_CXX` variables to begin with. 

Going by the principle of least surprise, I'd expect to be able to set the C++ flags by setting `CMAKE_CXX_FLAGS` from the outside.

How about checking if our forced `extra_flags` are already present in `CMAKE_CXX_FLAGS`, and if not appending them? This would avoid the cache cycle problem, and allow the flags to be set at the command line or in presets.

Reply:
Like we do in every R&D project (and in atlasexternals)?
  - https://github.com/acts-project/vecmem/blob/main/cmake/vecmem-functions.cmake#L140-L160
  - https://github.com/acts-project/detray/blob/main/cmake/detray-functions.cmake#L194-L214
  - ...
  - https://gitlab.cern.ch/atlas/atlasexternals/-/blob/main/Build/AtlasCMake/modules/AtlasCompilerSettings.cmake?ref_type=heads#L83-101

Yeah, I think that's a fine approach. :laughing:

Reply:
Hm, yeah but that comes at the cost of using a bunch of user functions, which to me is an antipattern.
I'll stick to the list mechanism and removal of duplicates.

Reply:
I've tested this locally, and there's no cache retention problem (`CMAKE_CXX_FLAGS`) is not a cache variable. You can also set it just fine from an enclosing project, so I think it checks all of our boxes.

Reply:
To be more correct, `CMAKE_CXX_FLAGS` **is** a cache variable "in general", but since in this directory you set it as a regular variable, for Acts it becomes a regular variable. :thinking:

So okay, that should generally be fine then. :thinking: 

But the "bunch of user functions"... I think it's much worse of a pattern that Acts doesn't provide any helper functions for building its libraries/executables. This means that any systematic change has to be done in N places. Making it very easy to arrive at a setup where some subdirectories do thing in one way, while others in some different one. :frowning: I myself am a **big** fan of the project-specific helper functions. (Athena would've drowned already without the `atlas_<foo>(...)` functions.)

Reply:
I think we'll have to agree to disagree on this.

The custom functions are a big contributing factor to most people not really knowing what's going on in the CMake configuration, since everything is hidden behind magic functions. For example, we can probably count the number of people who understand how packages **actually** work on one hand. 

I do concede, however, that the ATLAS build is a different beast. 

For ACTS (and imo also for the R&D projects) I think we're much better off without various opaque layers of CMake scripting.

Reply:
You know that I personally like CMake to a reasonable amount. But even I don't think that everybody would need to know as much as I do, just to develop some C\+\+ code for Acts. :thinking: Not everybody has to know everything. That's the whole point of using "functions" in the first place, in any language.

So yeah, I'm on a very different opinion.